### PR TITLE
Suggest `impl Iterator` when possible for `_` return type

### PR DESCRIPTION
**Pull request by estebank**
Address #106096.

### CODE REVIEW

File: compiler/rustc_hir_analysis/src/collect.rs [hunk 1]
```diff
@@ -1242,6 +1258,52 @@ fn infer_return_ty_for_fn_sig<'tcx>(
...
+    span: Span,
+    hir_id: hir::HirId,
+    def_id: LocalDefId,
+) -> Option<Ty<'tcx>> {
+    let Some(iter_trait) = tcx.get_diagnostic_item(sym::Iterator) else { return None; };
+    let Some(iterator_item) = tcx.get_diagnostic_item(sym::IteratorItem) else { return None; };
+    if !tcx
+        .infer_ctxt()
+        .build()
+        .type_implements_trait(iter_trait, [ret_ty], tcx.param_env(iter_trait))
```
Comment (compiler-errors) [inline compiler/rustc_hir_analysis/src/collect.rs:1273]: This should be checked in the param-env of the function, not in the param-env of the `Iterator` trait
```suggestion
        .type_implements_trait(iter_trait, [ret_ty], tcx.param_env(def_id))
```

File: compiler/rustc_hir_analysis/src/collect.rs [hunk 2]
```diff
@@ -1242,6 +1258,52 @@ fn infer_return_ty_for_fn_sig<'tcx>(
...
+        return None;
+    }
+    let infcx = tcx.infer_ctxt().build();
+    let ocx = ObligationCtxt::new_in_snapshot(&infcx);
+    // Find the type of `Iterator::Item`.
+    let origin = TypeVariableOrigin { kind: TypeVariableOriginKind::TypeInference, span };
+    let ty_var = infcx.next_ty_var(origin);
+    let projection = ty::Binder::dummy(ty::PredicateKind::Clause(ty::Clause::Projection(
+        ty::ProjectionPredicate {
+            projection_ty: tcx.mk_alias_ty(iterator_item, tcx.mk_substs([ret_ty.into()].iter())),
```
Comment (compiler-errors) [inline compiler/rustc_hir_analysis/src/collect.rs:1285]: nit: `mk_alias_ty` takes an `impl IntoIterator<Output = impl Into<GenericArg<'tcx>>>` -- I think this works instead?

```suggestion
            projection_ty: tcx.mk_alias_ty(iterator_item, [ret_ty]),
```

File: compiler/rustc_hir_analysis/src/collect.rs [hunk 3]
```diff
@@ -1242,6 +1258,52 @@ fn infer_return_ty_for_fn_sig<'tcx>(
...
+        },
+    )));
+    // Add `<ret_ty as Iterator>::Item = _` obligation.
+    ocx.register_obligation(crate::traits::Obligation::misc(
+        tcx,
+        span,
+        hir_id,
+        tcx.param_env(def_id),
+        projection,
+    ));
```
Comment (compiler-errors) [inline compiler/rustc_hir_analysis/src/collect.rs:1296]: I think you can just construct a projection type then pass it to `ocx.normalize` instead, should make the code a lot simpler

File: compiler/rustc_hir_analysis/src/collect.rs [hunk 4]
```diff
@@ -1242,6 +1258,52 @@ fn infer_return_ty_for_fn_sig<'tcx>(
...
+    if ocx.select_where_possible().is_empty()
+        && let item_ty = infcx.resolve_vars_if_possible(ty_var)
+        && !item_ty.references_error()
```
Comment (compiler-errors) [inline compiler/rustc_hir_analysis/src/collect.rs:1299]: Might be worth checking [`is_suggestable`](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_middle/ty/trait.IsSuggestable.html#tymethod.is_suggestable) so we don't suggest something like `impl Iterator<Output = [closure]>`

**Review by compiler-errors — COMMENTED** [review]
r=me with nits or not

File: src/test/ui/typeck/typeck_type_placeholder_item.stderr [hunk 5]
```diff
@@ -428,6 +428,27 @@ LL | const _: Option<_> = map(value);
...
+   |                               not allowed in type signatures
+   |                               help: replace with an appropriate return type: `impl Iterator<Item = usize>`
+
+error[E0121]: the placeholder `_` is not allowed within types on item signatures for constants
+  --> $DIR/typeck_type_placeholder_item.rs:229:10
+   |
+LL | const _: _ = (1..10).filter(|x| x % 2 == 0).map(|x| x * x);
+   |          ^ not allowed in type signatures
+   |
+note: however, the inferred type `Map<Filter<std::ops::Range<i32>, [closure@$DIR/typeck_type_placeholder_item.rs:229:29: 229:32]>, [closure@$DIR/typeck_type_placeholder_item.rs:229:49: 229:52]>` cannot be named
```
Comment (scottmcm) [inline src/test/ui/typeck/typeck_type_placeholder_item.stderr:446]: Ooh, this note is great!  Thanks!

File: compiler/rustc_hir_analysis/src/collect.rs (hunk 3 above)
Comment (compiler-errors) [reply compiler/rustc_hir_analysis/src/collect.rs:1294]: I guess I'll fix this in a follow-up PR

**Comment by compiler-errors** [comment]
@bors r-

tests needs bless :smile_cat:

File: compiler/rustc_hir_analysis/src/collect.rs (hunk 3 above)
Comment (estebank) [reply compiler/rustc_hir_analysis/src/collect.rs:1294]: I tried making that change, but I must have done it incorrectly and it didn't work. Let's land this and clean it up soon.

Comment (compiler-errors) [reply compiler/rustc_hir_analysis/src/collect.rs:1294]: Don't worry about it at all, I got it working on a local branch in any case. I do wonder what was wrong, though.

**Pull request merged**